started adding rules for more clock town locations

RevelationOrange opened the pull request:
also tweaked a few existing rules

RevelationOrange committed:
started adding rules for more clock town locations
also tweaked a few existing rules

Vlix commented on Rules.py:
```diff
@@ -51,12 +51,12 @@ def item_name(state, location):
-    set_rule(world.get_location('Great Bay GF Reward'), lambda state: state.has('GB SF'), 15)
-    set_rule(world.get_location('Stone Tower GF Reward'), lambda state: state.has('ST SF'), 15)
-    set_rule(world.get_location('Bomber Code'), lambda state: state.has('Magic Meter'))
+    set_rule(world.get_location('Clock Town GF Reward'), lambda state: state.has('CT SF', 1))
```
The default is already `1`, so it's superfluous, but maybe better for consistency with the faeries.

Vlix commented on Rules.py:
```diff
@@ -66,5 +66,62 @@ def global_rules(world):
     set_rule(world.get_location('Remove the Cursed Mask'), lambda state: state.has('Ocarina of Time'))
-    set_rule(world.get_location('Tunnel Balloon From Observatory'), lambda state: state.has('Bow') or (state.form('Deku') and state.has('Magic Meter')))
-    set_rule(world.get_location('Tunnel Balloon From ECT'), lambda state: state.has('Bow') or (state.form('Deku') and state.has('Magic Meter')))
+    set_rule(world.get_location('Tunnel Balloon From Observatory'), lambda state: (state.has('Bow') and state.form('Human')) or (state.form('Deku') and state.has('Magic Meter')))
```
I believe Zora Link can also pop the balloon with regular `B`

Vlix replied:
And I wonder if the Bow is the only item Human Link can use to pop balloons?

Vlix commented on Rules.py:
```diff
@@ -66,5 +66,62 @@ def global_rules(world):
+    # item names used are also guesses, as are some state function names probly
+    # location names may be longer than necessary so as to be descriptive, like we can change 'South Clock Town Hookshot
+    # Ledge Rupee Chest' if we want lol
+    set_rule(world.get_location('Clock Tower Platform HP'), lambda state: state.form('Human') or state.form('Goron') or state.form('Zora') or state.can_reach(world.get_location('Clock Town Business Scrub')))
```
Goron Link can't make this regularly, so it should be predicated on a trick, or not included.

Vlix replied:
`can_reach` is fine, but you also need Deku Link then.

Vlix commented on Rules.py:
```diff
@@ -66,5 +66,62 @@ def global_rules(world):
+    # location names may be longer than necessary so as to be descriptive, like we can change 'South Clock Town Hookshot
+    # Ledge Rupee Chest' if we want lol
+    set_rule(world.get_location('Clock Tower Platform HP'), lambda state: state.form('Human') or state.form('Goron') or state.form('Zora') or state.can_reach(world.get_location('Clock Town Business Scrub')))
+    set_rule(world.get_location('Festival Tower Rupee Chest'), lambda state: state.has('Hookshot') or state.can_reach(world.get_location('Clock Town Business Scrub')))
```
Hookshot also needs Human Link

Vlix commented on Rules.py:
```diff
@@ -66,5 +66,62 @@ def global_rules(world):
+    # I don't know that this item matters at all or if we should even have a spot for it
+    # but we can if we want, you get it the same way you get the code in the first cycle, you just have to be human to
+    # get the actual notebook from them I think
+    set_rule(world.get_location('Bomber Notebook'), lambda state: state.has('Magic Meter') and state.form('Deku') and state.form('Human'))
```
Human Link is the only hard requirement, the other is `can_reach('pop NCT balloon')` or something.

Vlix commented on Rules.py:
```diff
@@ -66,5 +66,62 @@ def global_rules(world):
+    # since getting the blast mask just involves slashing sakon, it might be possible to get it with other forms? like
+    # maybe zora? probly not goron though
+    # also you might need to be link to talk to the lady after, so maybe it is only human
+    set_rule(world.get_location('Blast Mask'), lambda state: state.form('Human'))
```
Zora Link can also get it (I think?) it's in my Google Sheet

Vlix commented on Rules.py:
```diff
@@ -66,5 +66,62 @@ def global_rules(world):
+    # also you might need to be link to talk to the lady after, so maybe it is only human
+    set_rule(world.get_location('Blast Mask'), lambda state: state.form('Human'))
+
+    set_rule(world.get_location('North Clock Town Tree HP'), lambda state: True)
```
Some forms can't get this, see the Google Sheet

Vlix commented on Rules.py:
```diff
@@ -66,5 +66,62 @@ def global_rules(world):
+
+    # I need to figure out all the actual requirements for this, including tricks, so this one is tentative
+    # also the trick names are guesses for sure
+    set_rule(world.get_location('East Clock Town 100 Rupee Chest'), lambda state: state.form('Human') or state.can('Goron Boost') or state.can('Gainer'))
```
Zora Link can also easily make this

Vlix commented on Rules.py:
```diff
@@ -66,5 +66,62 @@ def global_rules(world):
+    # also the trick names are guesses for sure
+    set_rule(world.get_location('East Clock Town 100 Rupee Chest'), lambda state: state.form('Human') or state.can('Goron Boost') or state.can('Gainer'))
+
+    set_rule(world.get_location('Ocarina of Time'), lambda state: (state.form('Deku') and state.has('Magic Meter')) or (state.has('Bow') and state.form('Human')))
```
Zora Link can also easily get this

Vlix commented on Rules.py:
```diff
@@ -66,5 +66,62 @@ def global_rules(world):
+
+    set_rule(world.get_location('Ocarina of Time'), lambda state: (state.form('Deku') and state.has('Magic Meter')) or (state.has('Bow') and state.form('Human')))
+    set_rule(world.get_location('All Night mask'), lambda state: state.form('Human') and state.has('Giants Wallet'))
+    set_rule(world.get_location('Bigger Bomb Bag'), lambda state: state.has('Giants Wallet'))
```
And `has('Bomb Mask')`, since you have to AT LEAST once get it to get the Bigger Bomb Bag into the shop.

Vlix commented on Rules.py:
```diff
@@ -66,5 +66,62 @@ def global_rules(world):
+    set_rule(world.get_location('Bigger Bomb Bag'), lambda state: state.has('Giants Wallet'))
+
+    # you might not need to be human to get this, but I'd bet the shop owner won't sell to other forms, at least deku
+    set_rule(world.get_location('Bomb Bag'), lambda state: state.form('Human'))
```
Everyone can get this afaik

Vlix commented on Rules.py:
```diff
@@ -66,5 +66,62 @@ def global_rules(world):
+    set_rule(world.get_location('Sword School HP'), lambda state: state.form('Human'))
+
+    # i swear this should be an optional trick, it's so hard without the bunny hood lol
+    set_rule(world.get_location('Postman Game HP'), lambda state: state.can('Time ten') or state.has('Bunny Hood'))
```
Anyone can get this at any time. I do it in one try most of the time. There are technically no requirements

Vlix commented on Rules.py:
```diff
@@ -66,5 +66,62 @@ def global_rules(world):
+    set_rule(world.get_location('Deku Scrub Playground HP'), lambda state: state.form('Deku'))
+    set_rule(world.get_location('Great Fairy Mask'), lambda state: True)
+    set_rule(world.get_location('Clock Town Maze Minigame HP'), lambda state: state.form('Goron'))
+    set_rule(world.get_location('Bomber Hideout 100 Rupee Chest'), lambda state: state.has('Bomber Code') and (state.has('Bomb Bag') or state.has('Blast Mask')))
```
You technically don't need the Bomber Code, only access to the Bomber Tunnel/Astral Observatory. (and you can get in from Termina Field under certain conditions)

Vlix commented:
You did it! I've given some comments, as well.
Do take a look at the Google Sheet here though, lots of Clock Town things are already tested with forms and extra requirements:
[MM Randomizer Google Spreadsheet](https://docs.google.com/spreadsheets/d/1x2cAcTJwMIl0QkHMzm3ea7z-WI98KofR9Msz8LKYCrk/edit?usp=sharing)

Vlix reviewed (changes requested):
You did it! I've given some comments, as well.
Do take a look at the Google Sheet here though, lots of Clock Town things are already tested with forms and extra requirements:
[MM Randomizer Google Spreadsheet](https://docs.google.com/spreadsheets/d/1x2cAcTJwMIl0QkHMzm3ea7z-WI98KofR9Msz8LKYCrk/edit?usp=sharing)

RevelationOrange committed:
did some fixes @Velix requested
I totally forgot about the zora ability to throw blades
(unless that needs magic? I forget, but I don't think so)
gotta test a few of them, like the clock tower platform

RevelationOrange committed:
fixing up clock town checks and making sure all items are accounted for
added some functions to BaseClasses (has_paper and can_pop_balloon)
to make some checks easier to read

Vlix commented on Rules.py:
```diff
@@ -56,50 +56,89 @@ def global_rules(world):
-    # essentially does the same thing
+    set_rule(world.get_location('Bomber Notebook'), lambda state: state.can_pop_balloon() and state.form('Human'))
+
+    set_rule(world.get_location('Bomber Code'), lambda state: state.can_pop_balloon())
```
Only Human and Deku link can get the code, since they're racist towards Goron and Zora link

Vlix commented on Rules.py:
```diff
@@ -110,35 +149,78 @@ def global_rules(world):
-    set_rule(world.get_location('Tingle Woodfall Map'), lambda state: True)
-    set_rule(world.get_location('Keaton HP'), lambda state: state.has('Keaton Mask'))
+    set_rule(world.get_location('Tingle Clock Town Map'), lambda state: state.can_pop_balloon())
+    set_rule(world.get_location('Tingle Woodfall Map'), lambda state: state.can_pop_balloon())
```
Both of these are just `True`, since Tingle isn't always up in the air.

Vlix commented on Rules.py:
```diff
@@ -110,35 +149,78 @@ def global_rules(world):
-    # you might not need to be human to get this, but I'd bet the shop owner won't sell to other forms, at least deku
-    set_rule(world.get_location('Bomb Bag'), lambda state: True)
-    # this can be tested easily
+    set_rule(world.get_location('Magic from NCT Great Fairy'), lambda state: state.form('Deku'))
```
This works for any form

Vlix commented on Rules.py:
```diff
@@ -110,35 +149,78 @@ def global_rules(world):
+    set_rule(world.get_location('Clock Town Shooting Gallery Quiver Prize'), lambda state: state.has('Bow') and state.form('Human'))
+    set_rule(world.get_location('Clock Town Shooting Gallery HP Prize'), lambda state: state.has('Bow') and state.form('Human'))
+
+    set_rule(world.get_location('Kafei Mask'), lambda state: state.form('Human'))
```
Any form but Deku will get you the Kafei mask. (it's in the second tab of the Google sheet, anything with an X of Check was tested explicitly by me)

Vlix reviewed (changes requested):
Ok, nice changes. More comments make it better to later on know if we need to test certain things.
It's a good idea to, when adding code, give comments on what is tentative and whatnot.
`TODO:`s make for easy checking what still needs to be changed/tested.

Fix these last little things and we can merge these changes. 👍

RevelationOrange committed:
more fixes for clock town stuff
commented out open checks that were just 'True'
I THINK clock town is complete now?
unless there are some obscure chests or something I missed
